### PR TITLE
trace/traceutil: introduce RoundTrip

### PR DESCRIPTION
**Pull request by bmizerany**
The purpose of this PR is to introduce/propose and discuss a new subpackage traceutil for housing various tracing utility functions for use with the trace package. It is intended that utility functions in this package be "boring" in that they are common patterns for tracing Go things.

The first of these "boring" utility function proposals is RoundTrip. RoundTrip wraps a http.RoundTripper with spans using the httptrace package and produces a result suitable for a FinishFunc of which there are two pre-made. One for a basic span, and another for detailed spans.

I'm very happy to discuss the design and further explain my motivation here if anyone has questions about anything that may be unclear.

-b

### CODE REVIEW

File: traceutil/roundtrip.go [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+	// TODO(bmizerany): check if already configured for propagation as to
+	// not override anything custom? We could do this with a key in the
+	// req.Context.
+	propagateHTTP(req)
+
+	trace.Request = req
+	trace.Started = timeNow()
+
+	res, err := r.inner.RoundTrip(req)
+	if err != nil {
```
Comment (rakyll) [inline traceutil/roundtrip.go:93]: If there is an error, shouldn't we still trace this request? Setting the Err field?

File: traceutil/roundtrip.go [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+//
+func FinishDetail() FinishFunc {
+	panic("TODO")
+}
+
+func (r traceRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	var (
+		trace   = newClientTrace()
+		ctx     = httptrace.WithClientTrace(req.Context(), trace.client)
+		timeNow = timeNowFromContext(ctx)
```
Comment (rakyll) [inline traceutil/roundtrip.go:80]: I'd prefer you to use timeNowFromContext everywhere. Req.Context is abusing the context package by exporting the context as a field (not to break go1 compat promise), but you'd expect functions working with ctx is not preserving the ctx state.

**Comment by rakyll** [comment]
Wait, I am confused. Shouldn't this package provide a RoundTripper to create github.com/rakyll/trace spans? I don't see it creating any spans but reporting ClientTrace stuff with a different API.

**Comment by bmizerany** [comment]
That is what the FinishFuncs do. Make sense?

**Comment by rakyll** [comment]
That makes total sense now! I will take a look at the code soon and will extend the trace library with a constructor, so you can create a Span with a specific start and end time.

File: traceutil/roundtrip.go [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+type RoundTripResult struct {
+	Started              time.Time
+	WroteHeaders         time.Time
+	WroteRequest         time.Time
+	GotFirstResponseByte time.Time
+	ReadHeaders          time.Time
+	ReadResponse         time.Time
+
+	BytesWritten int // TODO: implemenent
+	BytesRead    int
```
Comment (rakyll) [inline traceutil/roundtrip.go:23]: These two can be excessive values, right? s/int/int64.

File: traceutil/roundtrip.go [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+func propagateHTTP(req *http.Request) {
+	// TODO: implement
+}
+
+type traceRoundTripper struct {
+	inner    http.RoundTripper
+	finished FinishFunc
+}
+
+// RoundTrip returns a http.RoundTripper that wraps inner for purposes of
```
Comment (rakyll) [inline traceutil/roundtrip.go:45]: wraps the inner

File: traceutil/roundtrip.go [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+// completion or error of the returned http.RoundTripper.
+func RoundTrip(inner http.RoundTripper, finished FinishFunc) http.RoundTripper {
+	return traceRoundTripper{inner: inner, finished: finished}
+}
+
+type FinishFunc func(result *RoundTripResult)
+
+// FinishBasic records a single span for the request representing the from
+// first request byte written to last response byte read.
+func FinishBasic() FinishFunc {
```
Comment (rakyll) [inline traceutil/roundtrip.go:56]: s/FinishBasic/Finish

Let's keep the naming simple, any additional finishers can extend the name.

File: traceutil/roundtrip.go [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+//    |                                  ==ReadFirstByte==               |
+//    |                                                   ==ReadHeaders==|
+//    |==================================================================|
+//
+func FinishDetail() FinishFunc {
+	panic("TODO")
+}
+
+func (r traceRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	var (
```
Comment (rakyll) [inline traceutil/roundtrip.go:77]: Why not just

```
trace := newClientTrace()
ctx := httptrace.WithClientTrace(req.Context(), trace.client)
...
```
Then, we can have less indented code?

File: traceutil/roundtrip_test.go [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+		b.ReportAllocs()
+
+		r := RoundTrip(inner, nil)
+		for i := 0; i < b.N; i++ {
+			innerRes.Body = body
+
+			res, err := r.RoundTrip(new(http.Request))
+			if err != nil {
+				panic(err)
```
Comment (rakyll) [inline traceutil/roundtrip_test.go:43]: t.Fatal(err)

File: traceutil/roundtrip_test.go [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+	"testing"
+)
+
+type roundTripperFunc func(*http.Request) (*http.Response, error)
+
+func (f roundTripperFunc) RoundTrip(req *http.Request) (*http.Response, error) { return f(req) }
+
+func BenchmarkRoundTrip(b *testing.B) {
+	b.Run("basic", func(b *testing.B) {
+		var (
```
Comment (rakyll) [inline traceutil/roundtrip_test.go:17]: remove var, save from extra indentation.

File: traceutil/roundtrip_test.go [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+				})
+			)
+
+			r := RoundTrip(inner, nil)
+			for pb.Next() {
+				innerRes.Body = body
+
+				res, err := r.RoundTrip(new(http.Request))
+				if err != nil {
+					panic(err)
```
Comment (rakyll) [inline traceutil/roundtrip_test.go:76]: ditto

File: traceutil/roundtrip_test.go [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+				panic(err)
+			}
+			res.Body.Close()
+		}
+	})
+
+	b.Run("parallel", func(b *testing.B) {
+		b.ReportAllocs()
+		b.RunParallel(func(pb *testing.PB) {
+			var (
```
Comment (rakyll) [inline traceutil/roundtrip_test.go:52]: ditto

File: traceutil/roundtrip_test.go [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+	"net/http/httptrace"
+	"testing"
+)
+
+type roundTripperFunc func(*http.Request) (*http.Response, error)
+
+func (f roundTripperFunc) RoundTrip(req *http.Request) (*http.Response, error) { return f(req) }
+
+func BenchmarkRoundTrip(b *testing.B) {
+	b.Run("basic", func(b *testing.B) {
```
Comment (rakyll) [inline traceutil/roundtrip_test.go:16]: I'd prefer to have different Benchmark functions written for basic and parallel.

Sub-tests and benchmarks are useful if you share code in between test cases. In our case, there is no code is shared and having sub-benchs makes the code quite indented.

File: traceutil/roundtrip.go [hunk 12]
```diff
@@ -0,0 +1,191 @@
+package traceutil
```
Comment (rakyll) [inline traceutil/roundtrip.go:1]: Add a few words documenting traceutil package.

File: traceutil/roundtrip.go [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+}
+
+func (t traceBody) Read(b []byte) (int, error) {
+	n, err := t.body.Read(b)
+	t.result.BytesRead += n
+	if err != nil {
+		if err == io.EOF {
+			err = nil
+		}
+		t.finish(err)
```
Comment (rakyll) [inline traceutil/roundtrip.go:133]: We don't finish the spans until the body is entirely read.

What would happen, if user just wants to make a HEAD request and never cares about reading the body?

I think we should create and manage spans as things happen rather than delegating this work to the finish functions. It will make instrumentation more self-contained and each aspect will be independently traced rather than waiting for a finisher to do all the job.

**Review by rakyll — COMMENTED** [review]
Thanks! I left some comments.

I would prefer us to create spans as we need rather than delegating all the work to the finisher. It makes it hard to follow the execution and figure out cases where finisher will never be called. We are also interested in cases where our programs starts a request but crashes constantly without having a response back from the server. We should mark and label spans as soon as possible.

File: traceutil/roundtrip.go (hunk 3 above)
Comment (bmizerany) [reply traceutil/roundtrip.go:23]: Noted.

File: traceutil/roundtrip.go (hunk 5 above)
Comment (bmizerany) [reply traceutil/roundtrip.go:56]: Agreed.

File: traceutil/roundtrip.go (hunk 6 above)
Comment (bmizerany) [reply traceutil/roundtrip.go:77]: Okay.

File: traceutil/roundtrip.go (hunk 2 above)
Comment (bmizerany) [reply traceutil/roundtrip.go:80]: Okay

File: traceutil/roundtrip.go (hunk 1 above)
Comment (bmizerany) [reply traceutil/roundtrip.go:93]: I had a note for that and forgot to submit the review. I'll reaudit the code soon. I just wanted to get the basics out there to discuss API.

File: traceutil/roundtrip.go (hunk 13 above)
Comment (bmizerany) [reply traceutil/roundtrip.go:133]: > We don't finish the spans until the body is entirely read.

Not true. We finish on Close as well and that is a required call by the user as per the docs in net/http.

> What would happen, if user just wants to make a HEAD request and never cares about reading the body?

They still have to close res.Body.

> I think we should create and manage spans as things happen rather than delegating this work to the finish functions. It will make instrumentation more self-contained and each aspect will be independently traced rather than waiting for a finisher to do all the job.

I've addressed this here https://github.com/rakyll/trace/pull/4#issuecomment-282440824

File: traceutil/roundtrip_test.go (hunk 11 above)
Comment (bmizerany) [reply traceutil/roundtrip_test.go:16]: Okay.

File: traceutil/roundtrip_test.go (hunk 8 above)
Comment (bmizerany) [reply traceutil/roundtrip_test.go:17]: okay.

File: traceutil/roundtrip_test.go (hunk 7 above)
Comment (bmizerany) [reply traceutil/roundtrip_test.go:43]: This is a `b` so there is no Fatal

File: traceutil/roundtrip_test.go (hunk 10 above)
Comment (bmizerany) [reply traceutil/roundtrip_test.go:52]: ditto my comment.

File: traceutil/roundtrip_test.go (hunk 9 above)
Comment (bmizerany) [reply traceutil/roundtrip_test.go:76]: ditto my comment.
